relay-metrics/src/protocol.rs

```diff
@@ -0,0 +1,796 @@
+
+/// The [value](Metric::value) of a metric.
+///
+/// [Histograms](MetricType::Histogram) and [counters](MetricType::Counter) require numeric values
+/// which can either be integral or floating point. In contrast, [sets](MetricType::Set) and
+/// [gauges](MetricType::Gauge) can store any unique value including custom strings.
+#[derive(Clone, Debug, PartialEq, Deserialize, Serialize)]
+#[serde(untagged)]
+pub enum MetricValue {
+    /// A signed integral value.
+    Integer(i64),
+    /// A signed floating point value.
+    Float(f64),
+    /// A custom string value.
+    ///
+    /// This value cannot be used in numeric metrics.
+    Custom(String),
+    // TODO: Uuid(Uuid),
+}
+
+impl fmt::Display for MetricValue {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            MetricValue::Float(float) => float.fmt(f),
+            MetricValue::Integer(int) => int.fmt(f),
+            MetricValue::Custom(string) => string.fmt(f),
+        }
+    }
+}
+
+impl std::str::FromStr for MetricValue {
+    type Err = ParseMetricError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        Ok(if let Ok(int) = s.parse() {
+            Self::Integer(int)
+        } else if let Ok(float) = s.parse() {
+            Self::Float(float)
+        } else {
+            Self::Custom(s.to_owned())
+        })
+    }
+}
+
+/// The [type](Metric::ty) of a metric, determining its aggregation and evaluation.
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+pub enum MetricType {
+    /// Counts instances of an event.
+    ///
+    /// Counters can be incremented and decremented. The default operation is to increment a counter
+    /// by `1`, although increments by larger values are equally possible.
+    ///
+    /// This metric requires numeric values.
+    Counter,
+    /// Builds a statistical distribution over values reported.
+    ///
+    /// Based on individual reported values, histograms allow to query the maximum, minimum, or
+    /// average of the reported values, as well as statistical quantiles. With an increasing number
+    /// of values in the histogram, its accuracy becomes approximate.
+    Histogram,
+    /// Counts the number of unique reported values.
+    ///
+    /// Sets allow sending arbitrary discrete values and store the deduplicated count. With an
+    /// increasing number of unique values in the set, its accuracy becomes approximate. It is not
+    /// possible to query individual values from a set.
+    Set,
+    /// Stores absolute snapshots of values.
+    ///
+    /// Contrary to [counters](Self::Counter), which allow relative changes, gauges always store the
+    /// last absolute value submitted.
+    Gauge,
+}
+
+impl fmt::Display for MetricType {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            MetricType::Counter => f.write_str("c"),
+            MetricType::Histogram => f.write_str("h"),
```

Comment:
and prefer d over h

Reply:
Also a reply to the above comment: StatsD does not have a concept of a "distribution" type or "d", and "histograms" generally seems the more widespread term for this. So far, we're using histogram consistently except for one case where we use the natural language term distribution to explain what a histogram is.

Do you still prefer `d` and "distribution"?

Reply:
Did more research and realize that distribution may be the more accepted term. Updated everywhere.